plaid-stl/src/github/mod.rs

```diff
@@ -29,6 +29,99 @@ pub struct RepositoryCollaborator {
     pub permissions: Permission,
 }
 
+// Structs used for deserializing GH's responses when searching for code
+
+#[derive(Debug, Deserialize)]
+pub struct FileSearchResult {
+    pub total_count: u64,
+    pub incomplete_results: bool,
+    pub items: Vec<FileSearchResultItem>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct FileSearchResultItem {
+    pub path: String,
+    pub html_url: String,
+    pub url: String,
+    pub repository: GithubRepository,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GithubRepository {
+    pub name: String,
+    pub full_name: String,
+    pub private: bool,
+    pub description: Option<String>,
+    pub owner: GithubRepositoryOwner,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GithubRepositoryOwner {
+    pub login: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GithubFileContent {
+    #[serde(rename = "type")]
+    pub type_: String,
+    pub content: String,
+    pub encoding: String,
+}
+
+// END OF Structs used for deserializing GH's response when searching for code
+
+impl FileSearchResultItem {
+    /// Retrieve the content of the search result
+    pub fn retrieve_raw_content(&self) -> Result<String, PlaidFunctionError> {
+        let reference_regex = regex::Regex::new(r"^.*?ref=([a-f0-9]{40})$").unwrap(); // TODO improve
+        let reference = reference_regex
+            .captures(&self.url)
+            .ok_or(PlaidFunctionError::InternalApiError)?
+            .get(1)
+            .ok_or(PlaidFunctionError::InternalApiError)?
+            .as_str();
+        let content = fetch_file(
+            &self.repository.owner.login,
+            &self.repository.name,
+            &self.path,
+            reference,
+        )
+        .map_err(|_| PlaidFunctionError::InternalApiError)?;
+        let content = serde_json::from_str::<GithubFileContent>(&content)
+            .map_err(|_| PlaidFunctionError::InternalApiError)?;
+        if content.type_ != "file" || content.encoding != "base64" {
+            return Err(PlaidFunctionError::InternalApiError); // TODO not the right error
+        }
+        // base64 decode and return the corresponding string
+        Ok(String::from_utf8(
+            base64::decode(content.content.replace("\n", ""))
+                .map_err(|_| PlaidFunctionError::InternalApiError)?,
+        )
+        .map_err(|_| PlaidFunctionError::InternalApiError)?)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+/// Specifies criteria according to which results returned
+/// by the API should be included or discarded.
+pub struct CodeSearchCriteria {
+    /// Keep only results from this GH org
+    pub only_from_org: Option<String>,
+    /// Keep only results from these repositories (mutually exclusive with `discard_from_repos`)
+    pub only_from_repos: Option<Vec<String>>,
+    /// Discard files from these repositories (mutually exclusive with `only_from_repos`)
+    pub discard_from_repos: Option<Vec<String>>,
```

Comment:
Then this should be an enum so only one can be passed

Reply:
Done